File: 3967-earliest-finish-time-for-land-and-water-rides-ii/3967-earliest-finish-time-for-land-and-water-rides-ii.js
/**
 * @param {number[]} landStartTime
 * @param {number[]} landDuration
 * @param {number[]} waterStartTime
 * @param {number[]} waterDuration
 * @return {number}
 */
var earliestFinishTime = function(ls, ld, ws, wd) {
    const l = Math.min(...ls.map((s,i) => s + ld[i]));
    const w = Math.min(...ws.map((s,i) => s + wd[i]));
    return Math.min(
        ...ws.map((s,i) => Math.max(l, s) + wd[i]),
        ...ls.map((s,i) => Math.max(w, s) + ld[i])
    );
};

File: 3967-earliest-finish-time-for-land-and-water-rides-ii/3967-earliest-finish-time-for-land-and-water-rides-ii.py
class Solution:
    def earliestFinishTime(self, landStartTime: List[int], landDuration: List[int], waterStartTime: List[int], waterDuration: List[int]) -> int:
        l = min(s + d for s, d in zip(landStartTime, landDuration))
        w = min(s + d for s, d in zip(waterStartTime, waterDuration))
        L = min(
            max(l, s) + d
            for s, d in zip(waterStartTime, waterDuration)
        )
        W = min(
            max(w, s) + d
            for s, d in zip(landStartTime, landDuration)
        )

        return min(L, W)